### PR TITLE
Garbage Collector- validate that the written report DataFrame isn't empty

### LINKED ISSUE #4238: Running GC in dry run mode always fails
When running GC in dry-run mode it will always fail because it tries to write a DataFrame with a report of the actual deleted objects to the bucket, but this is not possible when the DataFrame is empty.
Exception:
`Cannot use all columns for partition columns`

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
Check that the written DataFrame is not empty before writing it, so that it won't throw an exception.

### How was this tested?
I ran the GC in dry-run mode against a repo with ~850K objects to be deleted. It used to fail on that repo but now it finishes successfully.

Fixes #4238

### CODE REVIEW

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 1]
```diff
@@ -340,24 +340,36 @@ object GarbageCollector {
...
-    removed
-      .withColumn("run_id", lit(runID))
-      .write
-      .partitionBy("run_id")
-      .mode(SaveMode.Overwrite)
-      .parquet(
-        concatToGCLogsPrefix(storageNSForHadoopFS, s"deleted_objects/$time/deleted.parquet")
-      )
-
-    spark.close()
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:360]: No longer closing?

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 2]
```diff
@@ -340,24 +340,36 @@ object GarbageCollector {
...
+    if(!removed.isEmpty) {
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:363]: But: this finishes running silently if it succeeded.  So any program downstream that needs to process the list of deleted objects now has to have a special-case for a GC run that succeeded without deleting anything.

I would very much prefer a consistent output, of an empty object.  There do exist empty Parquet files.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!  This definitely solves the issue, but it introduces an inconsistency that will be hard for automated programs to follow.

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 1 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:360]: closes in line 345

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 2 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala]: @arielshaqed

**Review by arielshaqed — APPROVED** [review]
Neat, thanks!
Mon blocking, but can the String be non-nullable? That will better reflect what it really is.

**Comment by arielshaqed** [comment]
Also, I think this is worth releasing today any which way. It's a serious noisy to usability!

**Pull request merged**